### DISCUSSION TITLE
How to avoid Cycle detected when publish event and listen result of this one

### DISCUSSION THREAD

**Discussion by thanhtrixx** [spring-projects/spring-modulith#921] [Q&A] [answered]
Hi team,

I want to implement an example of Coffee Shop. When an Order module is received order request, there will be a brewing event to the Brewing module and after completion, the results will be sent to the Order module. You can see the document generated below:
![components](https://github.com/user-attachments/assets/081fb484-59e8-4477-bfd0-0d6035b3a665)

Unfortunately when I verify:
```kotlin
  @Test
  fun verifiesModularStructure() {
    val modules = ApplicationModules.of(CoffeeShopApp::class.java)
    modules.verify()
  }
```
Then got the following error:
```
Cycle detected: Slice brewing -> 
                Slice order -> 
                Slice brewing
```
A more specific flow is as follows:
`OrderController -> OrderService (publish event) -> BrewingListener (publish event) -> OrderListener`

Module structure:
```
.
├── CoffeeshopApp.kt
├── brewing
│   ├── BrewingResultCommand.kt
│   └── internal
│       └── BrewingListener.kt
├── common
│   ├── Log.kt
│   ├── config
│   └── exception
│       ├── InvalidParamException.kt
│       └── ModuleMetadata.kt
└── order
    ├── BrewingCommand.kt
    ├── CreateOrderReqRes.kt
    ├── OrderController.kt
    ├── OrderInfo.kt
    ├── OrderService.kt
    └── internal
        ├── OrderEntity.kt
        ├── OrderListener.kt
        ├── OrderRepository.kt
        └── OtherServiceImpl.kt
```
Some other information:
```
Java: 21
Kotlin: 2.0.0
Spring Boot: 3.3.5
Spring Modulith: 1.2.4
```

I'm not sure if there's anything wrong here?

**Comment by jfoder** [comment]
You need to decide which module depends on which. In your case, it seems that the **Order** module is dependent on the **Brewing** module (**Order** requires actions from **Brewing** to execute a certain scenario). Therefore, in my opinion, it makes more sense for the **Brewing** module to expose the interface of events it can accept, while the **Order** module emits an event that implements this interface. This way, the **Brewing** module will depend on the interface it manages itself, eliminating the dependency on the **Order** module.

**Comment by thanhtrixx** [comment, marked as answer]
Finally, I found the solution. Just add metadata for `order` module to define it is a OPEN module and allow Cycle listen
You can find this commit: https://github.com/thanhtrixx/coffee-shop/commit/b083436f6d01844b71404a523e2a0f34cf99b52e
